src/Features/Core/Portable/GenerateMember/GenerateParameterizedMember/AbstractGenerateDeconstructMethodService.State.cs

```diff
@@ -0,0 +1,103 @@
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.GenerateMember.GenerateParameterizedMember
+{
+    internal partial class AbstractGenerateDeconstructMethodService<TService, TSimpleNameSyntax, TExpressionSyntax, TInvocationExpressionSyntax>
+    {
+        internal new class State :
+            AbstractGenerateParameterizedMemberService<TService, TSimpleNameSyntax, TExpressionSyntax, TInvocationExpressionSyntax>.State
+        {
+            /// <summary>
+            /// Make a State instance representing the Deconstruct method we want to generate.
+            /// The method will be called "Deconstruct". It will be a member of `typeToGenerateIn`.
+            /// Its arguments will be based on `targetVariables`.
+            /// </summary>
+            public static async Task<State> GenerateDeconstructMethodStateAsync(
+                TService service,
+                SemanticDocument document,
+                SyntaxNode targetVariables,
+                INamedTypeSymbol typeToGenerateIn,
+                CancellationToken cancellationToken)
+            {
+                var state = new State();
+                if (await state.TryInitializeMethodAsync(service, document, targetVariables, typeToGenerateIn, cancellationToken).ConfigureAwait(false))
+                {
+                    return state;
+                }
+
+                return null;
+            }
+
+            private async Task<bool> TryInitializeMethodAsync(
+                TService service,
+                SemanticDocument document,
+                SyntaxNode targetVariables,
+                INamedTypeSymbol typeToGenerateIn,
+                CancellationToken cancellationToken)
+            {
+                this.TypeToGenerateIn = typeToGenerateIn;
+                this.IsStatic = false;
+                this.IdentifierToken = service.MakeDeconstructToken();
+                this.MethodGenerationKind = MethodGenerationKind.Member;
+                MethodKind = MethodKind.Ordinary;
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var semanticModel = document.SemanticModel;
+                this.ContainingType = semanticModel.GetEnclosingNamedType(targetVariables.SpanStart, cancellationToken);
+                if (this.ContainingType == null)
+                {
+                    return false;
+                }
+
+                var parameters = TryMakeParameters(semanticModel, targetVariables, cancellationToken);
+                if (parameters.IsDefault)
+                {
+                    return false;
+                }
+
+                var methodSymbol = CodeGenerationSymbolFactory.CreateMethodSymbol(
+                    attributes: ImmutableArray<AttributeData>.Empty,
```

Comment:
can this not be 'default' as well?

Reply:
Yes. That works. It gets translated back to Empty down the chain of calls inside `CreateMethodSymbol`.

---
In reply to: [211398616](https://github.com/dotnet/roslyn/pull/28286#discussion_r211398616) [](ancestors = 211398616)